v8/third_party/heap/mmtk/mmtk.cc

```diff
@@ -176,17 +180,23 @@ bool Heap::IsValidHeapObject(HeapObject object) {
   return is_live_object(reinterpret_cast<void*>(object.address()));
 }
 
+// Access to the heap object iterator is assumed to be single-threaded.
+static bool iteration_finished = false;
+
 void Heap::ResetIterator() {
   TPHData* tph_data_ = get_tph_data(this);
   tph_archive_iter_reset(tph_data_->archive());
+  iteration_finished = false;
 }
 
 HeapObject Heap::NextObject() {
+  if (iteration_finished) return HeapObject();
   TPHData* tph_data_ = get_tph_data(this);
   void* obj_addr = tph_archive_iter_next(tph_data_->archive());
```

Comment:
I am slightly worried about the state `iteration_finished`. It would be good if this function can return null when the iteration finishes. For example, you can check for invalid values in `tph_archive_iter_next()` and return null.

Reply:
`iteration_finished` is removed.  I updated the code in `tph_archive_iter_next()`. See my explanation below.